rust/src/table/convert_to_delta.rs

```diff
@@ -0,0 +1,765 @@
+    delta_table(parquet_files, storage, partition_schema).await
+}
+
+/// Convert a Parquet table to a Delta table in place
+///
+/// # Arguments
+///
+/// * `path` - A path with "/" pointing at delta table root (i.e. where `_delta_log` is located).
+/// * `partition_schema`
+///     - A vector of [`SchemaField`] describing the data type, nullibility and metadata of partion columns.
+///     - If the schema of a partition column is not provided, schema data type is set to string and the field is set to nullable.
+///     - Errors if any partition column provided by the user does not exist in the parquet files.
+pub async fn convert_to_delta_from_path(
+    path: &str,
+    partition_schema: Vec<SchemaField>,
+) -> Result<DeltaTable, Error> {
+    match open_table(path).await {
+        Ok(table) => {
+            info!("A Delta table already exists in the given path");
+            Ok(table)
+        }
+        Err(_) => {
+            info!("Converting Parquet table in path: {path:?}");
+            let storage = object_store(path)?;
+            let parquet_files = parquet_files(storage.clone()).await?;
+            delta_table(parquet_files, storage, partition_schema).await
+        }
+    }
+}
+
+fn object_store(path: &str) -> Result<ObjectStoreRef, Error> {
+    debug!("Creating an object store for path: {path:?}");
+    let location = ensure_table_uri(path)?;
+    Ok(Arc::new(DeltaObjectStore::try_new(
+        location,
+        HashMap::new(),
+    )?))
+}
+
+async fn parquet_files(storage: ObjectStoreRef) -> Result<Vec<ObjectMeta>, Error> {
+    let objects = storage.list(None).await?;
+    let mut files = Vec::new();
+    let fut = objects.try_for_each_concurrent(10, |meta| {
+        if Some("parquet") == meta.location.extension() {
+            debug!("Found parquet file {:#?}", meta.location);
+            files.push(meta);
+        }
+        future::ready(Ok(()))
+    });
+    // Return immediately with an error if the stream returns an error
+    fut.await?;
+    if files.is_empty() {
+        return Err(Error::ParquetFileNotFound);
+    }
+    Ok(files)
+}
+
+async fn arrow_schema(file: ObjectMeta, storage: ObjectStoreRef) -> Result<ArrowSchema, Error> {
+    let mut arrow_schema =
+        ParquetRecordBatchStreamBuilder::new(ParquetObjectReader::new(storage, file))
+            .await?
+            .schema()
+            .as_ref()
+            .clone();
+    // Arrow schema of Parquet files may have conflicting metatdata. Since Arrow schema metadata is not used to generate Delta table schema, we set the metadata field to an empty HashMap
+    arrow_schema.metadata = HashMap::new();
+    Ok(arrow_schema)
+}
+
+fn add_action(
+    file: &ObjectMeta,
+    partition_values: HashMap<String, Option<String>>,
+) -> Result<Action, Error> {
+    Ok(Action::add(Add {
+        path: percent_decode_str(file.location.as_ref())
+            .decode_utf8()?
+            .to_string(),
```

Comment:
are we sure this location needs to be decoded before we pipe it through the add action?

Reply:
Percent decoding as UTF8 is needed here because the path of Add action is used to generate a `ScalarValue::Utf8` in the `build` method of `DeltaScanBuilder` https://github.com/delta-io/delta-rs/blob/main/crates/deltalake-core/src/delta_datafusion/mod.rs#L594-L595

Also, the `path` field has a [`#[serde(with = "serde_path")]` attribute](https://github.com/delta-io/delta-rs/blob/main/rust/src/protocol/mod.rs#L350-L351) applied to it. `path` will be decoded during [deserialization](https://github.com/delta-io/delta-rs/blob/main/rust/src/protocol/serde_path.rs#L11).